cpp/dolfinx/refinement/interval.h

```diff
@@ -0,0 +1,225 @@
+  std::int32_t number_of_refined_cells
+      = std::count(refinement_marker.begin(),
+                   std::next(refinement_marker.begin(),
+                             mesh.topology()->index_map(1)->size_local()),
+                   true);
+
+  // Produce local global indices, by padding out the previous index map
+  std::vector<std::int64_t> global_indices
+      = adjust_indices(*mesh.topology()->index_map(0), number_of_refined_cells);
+
+  // Build the topology on the new vertices
+  const auto refined_cell_count = mesh.topology()->index_map(1)->size_local()
+                                  + mesh.topology()->index_map(1)->num_ghosts()
+                                  + number_of_refined_cells;
+
+  std::vector<std::int64_t> cell_topology;
+  cell_topology.reserve(refined_cell_count * 2);
+
+  std::vector<std::int32_t> parent_cell;
+  parent_cell.reserve(refined_cell_count);
+
+  for (std::int32_t cell = 0; cell < map_c->size_local(); ++cell)
+  {
+    const auto& vertices = c_to_v->links(cell);
+    assert(vertices.size() == 2);
+
+    // we consider a (previous) cell, i.e. an edge of (global) vertices
+    // a ----------- b
+    const std::int64_t a = global_indices[vertices[0]];
+    const std::int64_t b = global_indices[vertices[1]];
+
+    if (refinement_marker[cell])
+    {
+      // find (global) index of new midpoint vertex:
+      // a --- c --- b
+      auto it = new_vertex_map.find(cell);
+      assert(it != new_vertex_map.end());
+      const std::int64_t c = it->second;
+
+      // add new cells/edges to refined topology
+      cell_topology.insert(cell_topology.end(), {a, c, c, b});
+      parent_cell.insert(parent_cell.end(), {cell, cell});
+    }
+    else
+    {
+      // copy the previous cell
+      cell_topology.insert(cell_topology.end(), {a, b});
+      parent_cell.push_back(cell);
+    }
+  }
+
+  assert(cell_topology.size() == refined_cell_count * 2);
+  assert(parent_cell.size() == refined_cell_count);
+
+  std::vector<std::int32_t> offsets(refined_cell_count + 1);
+  std::ranges::generate(offsets, [i = 0]() mutable { return 2 * i++; });
+
+  graph::AdjacencyList cell_adj(std::move(cell_topology), std::move(offsets));
+
+  return {std::move(cell_adj), std::move(new_vertex_coords), xshape,
+          std::move(parent_cell)};
+}
+
+} // namespace impl
+
+/// Refines a (topologically) one dimensional mesh by splitting cells, i.e.
+/// edges.
+///
+/// @param[in] mesh Mesh to be refined
+/// @param[in] cells Optional indices of the cells that should be split by this
+/// refinement. If not provided, all cells are considered marked for refinement,
+/// i.e. a uniform refinement is performed.
+/// @param[in] redistribute Option to enable redistribution of the refined mesh
+/// across processes.
+/// @param[in] ghost_mode Ghost mode of the refined mesh, default is ghost mode
+/// none
+///
+/// @return Refined mesh, and list of parent cells - for every new cell index
```

Comment:
and an array mapping the child cell index of the refined mesh to its parent cell index in the unrefined mesh.